### PR TITLE
Remove resize index settings once shards are started

### LINKED ISSUE #90127: Remove resize left-over at shard started
When a resize (shrink/clone/split) occurs, the new index gets `index.resize.source.uuid` and `index.resize.source.name` set in order to indicate the source of the operation. For shrink we also set `index.routing.allocation.initial_recovery.<something>` to force the location of the shrunk shards to go on the node holding the source shards.

Today this is left as is once the shard is started. But at this time, the shards are now decoupled and these indications are no longer used. This is enforced in allocation deciders by ensuring that we only check these for unassigned shards with recovery source `LOCAL_SHARDS`. Notice that the resize source index is often deleted afterwards, making the reference to the source index sort of dangling.

I propose to remove this at shard started (of the last shard), such that it is clear that these are no longer in use.

Relates #89768

### PR DESCRIPTION
**Pull request by tlrx**
This commit removes the specific index settings used to resize indices during clone/split/shrink.

Settings are removed once all primary shards are started. This is implemented in a new `RoutingChangesObserver` in order to not add more complexity to the existing `IndexMetadataUpdater`.

Closes #90127

Edit:  Sadly ILM relies on the `INDEX_RESIZE_SOURCE_NAME_KEY` setting to identify an index that has been shrunk. I suppose this information was not added to ILM lifecycle execution state to keep it minimal but it now makes this change harder to implement. I've updated the code to not remove this index settings for ILM managed indices but I would prefer ILM to maintain its own metadata in its execution state, specially when an information is required after the ILM action is done.

### CODE REVIEW

**Review by henningandersen — APPROVED** [review]
LGTM.

> ILM to maintain its own metadata in its execution state, specially when an information is required after the ILM action is done.

It looks like this is only used for safety checking, making me wonder if we can find another way to be equally safe. For instance we check that the source still exists in `CleanupShrinkIndexStep` but then fire a transport action to delete the target. There could be a race here in case we anticipate the source index might be deleted?

But happy to see this change, we can tackle ILM in a follow-up/next iteration.

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/ShrinkIndexIT.java [hunk 1]
```diff
@@ -688,5 +705,39 @@ public void testShrinkThenSplitWithFailedNode() throws Exception {
...
+        assertThat("Index " + index + " should have index metadata", indexMetadata, notNullValue());
+        assertThat("Index " + index + " should have index metadata", indexMetadata, notNullValue());
+        assertThat("Index " + index + " should not have resize source index", indexMetadata.getResizeSourceIndex(), nullValue());
+        assertThat(
+            "Index " + index + " should not have resize source UUID setting",
+            IndexMetadata.INDEX_RESIZE_SOURCE_UUID.exists(indexMetadata.getSettings()),
+            equalTo(false)
+        );
+        assertThat(
+            "Index " + index + " should not have resize source UUID setting",
```
[the remark this answers is not in the document]
Comment (tlrx) [reply server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/ShrinkIndexIT.java:733]: Thanks for spotting this, I pushed bdd2c2aff4837d860d309a92ca0c6a3e959d3944

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/ResizeSourceIndexSettingsUpdater.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            final Map<Index, Settings> updates = Maps.newHashMapWithExpectedSize(changes.size());
+            for (Index index : changes) {
+                var indexMetadata = metadata.getIndexSafe(index);
+                if (routingTable.index(index).allPrimaryShardsActive()) {
+                    assert indexMetadata.getResizeSourceIndex() != null : "no resize source index for " + index;
+
+                    Settings.Builder builder = Settings.builder().put(indexMetadata.getSettings());
+                    builder.remove(IndexMetadata.INDEX_SHRINK_INITIAL_RECOVERY_KEY);
+                    builder.remove(IndexMetadata.INDEX_RESIZE_SOURCE_UUID_KEY);
+                    if (Strings.isNullOrEmpty(indexMetadata.getLifecyclePolicyName())) {
```
[the remark this answers is not in the document]
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/ResizeSourceIndexSettingsUpdater.java:52]: ILM integ tests fail without this, but I added a random policy in the `ResizeSourceIndexSettingsUpdater` unit tests (5c362befb6f7bf6269421bf3fa6381fa7a57bfca).

**Comment by tlrx** [comment]
Thanks Henning. This is merged in 8.6.0 and I'll take a look at ILM after that.

**Pull request merged**